### PR TITLE
TandemFoil Paper: 3L/256d wider model at champion config

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The TFP champion is 3L/192d. Going deeper (4L/192d, PR #3094; 4L/256d, PR #3095) failed catastrophically — the pressure channel overflows via asinh transform. But going WIDER at the same depth has not been tested.

3L/256d keeps the proven 3L depth (which avoids the 4L instability) while adding 78% more capacity per layer (256^2 vs 192^2 in attention). This tests whether the model is width-limited at 192d. On DrivAerML, width scaling (256→320→384→512) was the single biggest improvement lever.

Note: 4L/256d failed with Infinity field_mse (PR #3095), but that was depth+width together. This isolates the width variable.

## Instructions

Run at the champion optimizer config with increased width:

**Run 1: 3L/256d/4H**
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset tandemfoil_paper \
  --optimizer lion \
  --lr 1.25e-4 \
  --cosine-t-max 10 \
  --grad-clip 0.5 \
  --weight-decay 1e-2 \
  --enable-fourier \
  --model-layers 3 \
  --model-hidden-dim 256 \
  --model-heads 4 \
  --enable-te-coord-frame \
  --enable-cp-panel \
  --enable-cp-panel-tandem-only \
  --asinh-pressure \
  --residual-prediction \
  --enable-pressure-prior-addition \
  --epochs 999 \
  --ema-decay 0.999 \
  --wandb_group tfp-width-scaling
```

Note: 4 heads at 256d = 64 dim/head (same as champion 3 heads at 192d = 64 dim/head). This keeps the attention head dimension constant while scaling width.

Report best val_primary/field_mse and divergence epoch. If the wider model increases per-epoch time significantly, also report epoch count reached vs champion.

## Baseline

- **val_primary/field_mse:** 0.002383 at epoch 443 (diverged ep462)
- **Config:** Lion lr=1.25e-4, T_max=10, gc=0.5, WD=1e-2, EMA=0.999, 3L/192d/3H, Fourier+physics
- **W&B run:** d1xh0o1p
- **Reproduce:** `cd target/icml2026 && python train.py --dataset tandemfoil_paper --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 0.5 --weight-decay 1e-2 --enable-fourier --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999 --ema-decay 0.999`

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT robin:

## Results

### TFP 3L/256d/4H at Champion Config — NEGATIVE

**Best val_primary/field_mse:** Infinity (best finite: 8.61e+24 at ep177) vs baseline 0.002383
**Divergence epoch:** 229 (vs champion 462)
**Epochs reached:** 279+ (still running, post-divergence)
**W&B run:** 3j33y6wi
**Peak memory:** 56.6 GB / 97.9 GB (GPU 0)

### Metrics Summary

| Metric | 3L/256d (this run) | 3L/192d (champion) |
|--------|-------------------|---------------------|
| field_mse | Infinity (best 8.61e+24 @ ep177) | 0.002383 @ ep443 |
| volume_mse | best 3.49e+22 @ ep220 | (included in field_mse) |
| surface_mse_Ux | 0.00113 | — |
| surface_mse_Uy | 0.00392 | — |
| volume_mse_Ux | 0.00438 | — |
| volume_mse_Uy | 0.00393 | — |
| Divergence | ep229 | ep462 |

### Command Used

```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset tandemfoil_paper --optimizer lion --lr 1.25e-4 --cosine-t-max 10 \
  --grad-clip 0.5 --weight-decay 1e-2 --enable-fourier \
  --model-layers 3 --model-hidden-dim 256 --model-heads 4 \
  --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only \
  --asinh-pressure --residual-prediction --enable-pressure-prior-addition \
  --epochs 999 --ema-decay 0.999 \
  --agent robin --wandb-name "robin/tfp-3l-256d-champion" --wandb-group tfp-width-scaling
```

### What Happened

**Definitive NEGATIVE result.** The 3L/256d wider model at champion config failed to produce finite field_mse and diverged at ep229.

**Training phases:**
1. **ep1–159:** field_mse=Infinity throughout. Loss declining (0.07→0.043), velocity components steadily improving. Volume and surface pressure stuck at Infinity due to sinh overflow in asinh pressure transform.
2. **ep160–228 (breakthrough window):** Pressure predictions started becoming finite. Best field_mse=8.61e+24 (ep177), best volume_mse=3.49e+22 (ep220). Velocity components reached excellent values (surface_Ux=0.00113, volume_Ux=0.00438). However, even the best field_mse was **27 orders of magnitude** worse than baseline (0.002383). Pressure oscillated wildly between finite (e+22 to e+30) and Infinity, driven by the T_max=10 cosine LR cycling.
3. **ep229+ (divergence):** Grad norms exploded from 8→53→124→269→429→1354→... Reached 658k at ep276. Loss went from 0.042 to >2.0. All velocity components destroyed (surface_Ux from 0.001 to >1.0). Model entered repeated divergence-partial recovery cycles but never recovered.

**Key observations:**
- The wider model (256d) was MORE stable than 4L/192d (diverged at ep229 vs ep116), confirming that width is safer than depth for this architecture
- Volume pressure was actively converging before divergence (3 consecutive improvements: ep206→ep219→ep220), suggesting the model was learning pressure but too slowly
- The fundamental blocker is the same as all other non-champion configurations: the sinh inverse transform (sinh(pred)/0.75) amplifies even slightly-off pressure predictions into astronomical or Infinity values
- Even at its best (ep177), the model was 8.61e+24 — this is not remotely competitive with 0.002383
- The champion 3L/192d somehow overcomes this pressure barrier (likely reaching a critical point around ep200-300 where pressure suddenly converges), but the wider 3L/256d model diverged before reaching that point

**Comparison with PR #3094 (4L/192d):**
- 4L/192d: Diverged ep116, field_mse=Infinity throughout
- 3L/256d: Diverged ep229, field_mse transiently finite (8.61e+24)
- Width is clearly safer than depth, and the wider model showed more pressure learning, but both failed to reach competitive pressure values

### Suggested Follow-ups

1. **Lower learning rate for wider model:** The divergence at ep229 may be LR-related. The wider model has 78% more parameters but uses the same lr=1.25e-4 tuned for 192d. Try lr=8e-5 or 1e-4 to give the pressure convergence more time.
2. **Longer T_max for wider model:** T_max=10 causes 36 full LR cycles per epoch, which was a clear source of pressure oscillation. T_max=50 or 100 would reduce the cycling amplitude and might allow pressure to converge instead of oscillating.
3. **Pressure gradient clipping:** The divergence cascaded through pressure. A separate, tighter gradient clip on the pressure channel (or clamping the sinh output) could prevent the catastrophic blow-up while allowing longer training.
4. **The champion's pressure convergence:** Understanding exactly how/when the 3L/192d champion's pressure becomes finite and converges to 0.002383 would inform all future scaling experiments. The W&B logs for d1xh0o1p should show this transition point.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3124 because ADVISOR: 3L/256d catastrophically diverges — field_mse 8.61e+24 at ep177, 27 orders of magnitude worse than baseline. Combined with #3157 (3L/224d, field_mse ~2.2e9), width beyond 192d is conclusively dead for TFP. The narrow stability window cannot support wider attention layers — both 224d and 256d amplify gradient instability fatally. 192d is confirmed as the capacity ceiling. Reassigning you to AirfRANS where we have a meaningful volume gap to close..
